docs/features/incremental-generators.cookbook.md

```diff
@@ -0,0 +1,810 @@
+    provider.GlobalOptions.TryGetValue("build_property.MyGenerator_EnableLogging", out var emitLoggingSwitch)
+        ? emitLoggingSwitch : false);
+```
+
+A user can thus enable, or disable logging, by setting a property in their project file.
+
+Now, consider that the generator author wants to optionally allow opting in/out of logging on a per-additional file basis. The author can request that MSBuild emit the value of metadata for the specified file, by adding to the `CompilerVisibleItemMetadata` item group. The author specifies both the MSBuild itemType they want to read the metadata from, in this case `AdditionalFiles`, and the name of the metadata that they want to retrieve for them.
+
+```xml
+<ItemGroup>
+    <CompilerVisibleItemMetadata Include="AdditionalFiles" MetadataName="MyGenerator_EnableLogging" />
+</ItemGroup>
+```
+
+This value of `MyGenerator_EnableLogging` will be emitted to a generated analyzer config file, for each of the additional files in the compilation, with an item name of `build_metadata.AdditionalFiles.MyGenerator_EnableLogging`. The generator can read this value in the context of each additional file:
+
+```cs
+context.AdditionalFilesProvider
+       .Combine(context.AnalyzerConfigOptionsProvider)
+       .Select((pair, ctx) =>
+           pair.Right.GetOptions(pair.Left).TryGetValue("build_metadata.AdditionalFiles.MyGenerator_EnableLogging", out var perFileLoggingSwitch)
+               ? perFileLoggingSwitch : false);
+```
+
+In the users project file, the user can now annotate the individual additional files to say whether or not they want to enable logging:
+
+```xml
+<ItemGroup>
+    <AdditionalFiles Include="file1.txt" />  <!-- logging will be controlled by default, or global value -->
+    <AdditionalFiles Include="file2.txt" MyGenerator_EnableLogging="true" />  <!-- always enable logging for this file -->
+    <AdditionalFiles Include="file3.txt" MyGenerator_EnableLogging="false" /> <!-- never enable logging for this file -->
+</ItemGroup>
+```
+
+**Full Example:**
+
+MyGenerator.props:
+
+```xml
+<Project>
+    <ItemGroup>
+        <CompilerVisibleProperty Include="MyGenerator_EnableLogging" />
+        <CompilerVisibleItemMetadata Include="AdditionalFiles" MetadataName="MyGenerator_EnableLogging" />
+    </ItemGroup>
+</Project>
+```
+
+MyGenerator.csproj:
+
+```xml
+<Project>
+  <PropertyGroup>
+    <GeneratePackageOnBuild>true</GeneratePackageOnBuild> <!-- Generates a package at build -->
+    <IncludeBuildOutput>false</IncludeBuildOutput> <!-- Do not include the generator as a lib dependency -->
+  </PropertyGroup>
+
+  <ItemGroup>
+    <!-- Package the generator in the analyzer directory of the nuget package -->
+    <None Include="$(OutputPath)\$(AssemblyName).dll" Pack="true" PackagePath="analyzers/dotnet/cs" Visible="false" />
+
+    <!-- Package the props file -->
+    <None Include="MyGenerator.props" Pack="true" PackagePath="build" Visible="false" />
```

Comment:
Maybe `buildTransitive`? Assuming that generators from NuGet packages flow transitively